src/naga/engine.clj

```diff
@@ -0,0 +1,85 @@
+(ns naga.engine
+  (:require [naga.structs :as st :refer [EPVPattern RulePatternPair Body Program]]
+            [naga.queue :as q]
+            [naga.store :as store]
+            [naga.util :as u]
+            [schema.core :as s])
+  (:import [naga.structs Rule]
+           [naga.store Storage]))
+
+(def true* (constantly true))
+
+(s/defn execute
```

Comment:
Can you break this up into multiple functions?  Its hard to follow the branching logic in the loop.